File: english-hub-backend/src/main/java/com/example/englishhubbackend/service/ExamService.java
package com.example.englishhubbackend.service;

import com.example.englishhubbackend.dto.request.*;
import com.example.englishhubbackend.dto.response.ExamResponse;
import com.example.englishhubbackend.dto.response.ExamSubmissionResponse;
import com.example.englishhubbackend.dto.response.QuestionGroupResponse;
import com.example.englishhubbackend.dto.response.QuestionResponse;
import java.util.List;
import java.util.UUID;

public interface ExamService {
  ExamResponse createExam(ExamCreateRequest examCreateRequest);

  List<ExamResponse> getAllExams();

  ExamResponse getExamById(UUID examId);

  ExamResponse updateExam(UUID examId, ExamUpdateRequest examUpdateRequest);

  void deleteExam(UUID examId);

  List<QuestionResponse> addQuestionsToExam(
      UUID examId, List<QuestionCreateRequest> questionCreateRequest);

  QuestionResponse addQuestionToExam(UUID examId, QuestionCreateRequest questionCreateRequest);

  List<QuestionResponse> getAllQuestionsFromExam(UUID examId);

  QuestionResponse updateQuestionInExam(
      UUID examId, UUID questionId, QuestionUpdateRequest questionUpdateRequest);

  List<QuestionGroupResponse> getQuestionGroupsFromExam(UUID examId);

  ExamSubmissionResponse submitExam(UUID examId, ExamSubmissionRequest examSubmissionRequest);

    void deleteQuestionFromExam(UUID examId, UUID questionId);
}

File: english-hub-frontend/src/services/examService.ts
import {
  ExamCreateRequest,
  ExamResponse,
  ExamUpdateRequest,
} from '@/types/examType';
import axiosInstance from './axiosInstance';
import { QuestionCreateRequest, QuestionResponse, QuestionUpdateRequest } from '@/types/questionType';

export async function getAllExams(): Promise<ExamResponse[]> {
  const response = await axiosInstance.get('/exams');
  return response.data.result;
}

export async function getExamById(examId: string): Promise<ExamResponse> {
  const response = await axiosInstance.get(`/exams/${examId}`);
  return response.data.result;
}

export async function createExam(
  exam: ExamCreateRequest
): Promise<ExamResponse> {
  const response = await axiosInstance.post('/exams', exam);
  return response.data.result;
}

export async function updateExam(
  examId: string,
  exam: ExamUpdateRequest
): Promise<ExamResponse> {
  const response = await axiosInstance.put(`/exams/${examId}`, exam);
  return response.data.result;
}

export async function deleteExam(examId: string): Promise<string> {
  const response = await axiosInstance.delete(`/exams/${examId}`);
  return response.data.message;
}

export async function addQuestionToExam(
  examId: string,
  question: QuestionCreateRequest
): Promise<QuestionResponse> {
  const response = await axiosInstance.post(
    `/exams/${examId}/question`,
    question
  );
  return response.data.result;
}

export async function updateQuestionInExam(
  examId: string,
  questionId: string,
  question: QuestionUpdateRequest
): Promise<QuestionResponse> {  
  const response = await axiosInstance.put(
    `/exams/${examId}/questions/${questionId}`,
    question
  );
  return response.data.result;
}

export async function addQuestionsToExam(
  examId: string,
  questions: QuestionCreateRequest[]
): Promise<QuestionResponse[]> {
  const response = await axiosInstance.post(
    `/exams/${examId}/questions`,
    questions
  );
  return response.data.result;
}

export async function getQuestionsFromExam(
  examId: string
): Promise<QuestionResponse[]> {
  const response = await axiosInstance.get(`/exams/${examId}/questions`);
  return response.data.result;
}

export async function getQuestionGroupsFromExam(
  examId: string
): Promise<QuestionResponse[]> {
  const response = await axiosInstance.get(`/exams/${examId}/questions/groups`);
  return response.data.result;
}
